gcsio/src/main/java/com/google/cloud/hadoop/gcsio/FileAccessPatternManager.java

```diff
@@ -151,18 +149,21 @@ private boolean isRandomAccessPattern(long currentPosition) {
     return false;
   }
 
+  private boolean isBackwardOrForwardSeekNotRequested() {
+    return !isBackwardSeekRequested && !isForwardSeekRequested;
+  }
+
   private boolean shouldDetectSequentialAccess() {
     return randomAccess
-        && !isBackwardSeekRequested
-        && !isForwardSeekRequested
+        && isBackwardOrForwardSeekNotRequested()
         && consecutiveSequentialCount >= readOptions.getFadviseRequestTrackCount()
         && readOptions.getFadvise() == Fadvise.AUTO_RANDOM;
   }
 
   private boolean shouldDetectRandomAccess() {
-    return !randomAccess
```

Comment:
If already in randomAccess mode there is no need to detect random access.

I do agree that the current logic will not update backwardSeek and forwardSeek if occurredwhen random access was set but I would suggest updating those flags well within fn:`updateAccessPattern` .